src/cloudant/_common_util.py

```diff
@@ -201,14 +202,16 @@ def py_to_couch_validate(key, val):
         if val not in ('ok', 'update_after'):
             raise CloudantArgumentError(135, val)
 
-def _py_to_couch_translate(key, val):
+def _py_to_couch_translate(key, val, encoder=None):
     """
     Performs the conversion of the Python parameter value to its CouchDB
     equivalent.
     """
     try:
         if key in ['keys', 'endkey_docid', 'startkey_docid', 'stale', 'update']:
             return {key: val}
+        if key in ['endkey', 'key', 'startkey'] and not isinstance(val, (int, LONGTYPE, bool)):
```

Comment:
`json.dumps` converts primitive types into strings, so the following assertions will fail if I remove the instance check (actual and expected values are swapped in our tests, so in these cases the actual value is the expected and vice versa):

https://github.com/cloudant/python-cloudant/blob/master/tests/unit/param_translation_tests.py#L47 with
```
AssertionError: {'endkey': '10'} != {'endkey': 10}
- {'endkey': '10'}
?            -  -

+ {'endkey': 10}
```

https://github.com/cloudant/python-cloudant/blob/master/tests/unit/param_translation_tests.py#L123 with

```
{'key': 10} != {'key': '10'}

Expected :{'key': '10'}
Actual   :{'key': 10}
```
and
https://github.com/cloudant/python-cloudant/blob/master/tests/unit/param_translation_tests.py#L197 with
```
{'startkey': 10} != {'startkey': '10'}

Expected :{'startkey': '10'}
Actual   :{'startkey': 10}
```

Reply:
>`json.dumps` converts primitive types into strings

Well `json.dumps` always produces a python string, that's what it's for, but it shouldn't produce a JSON string for a primitive type (see: https://docs.python.org/3/library/json.html#py-to-json-table)

So the behaviour of `json.dumps` is like this:
int: `10` -> `'10'` (i.e. a python string)
str: `'10'` -> `'"10"'` (i.e. a python string of a JSON string)

>the following assertions will fail if I remove the instance check

Based on above are the tests are wrong or do we encounter a second round of encoding somewhere before the query string?

Reply:
You are right, the tests are wrong. Fixed in a7ac8a7.

Reply:
OK, so I see the reason the tests passed previously was because the `TYPE_CONVERTERS` for numbers did
`lambda x: x` - and that in turn would pass the number type to the requests module.
AFAICT [requests will ensure a `str` param value is utf-8 encoded](https://github.com/psf/requests/blob/1466ad713cf84738cd28f1224a7ab4a19e50e361/requests/models.py#L104) but just pass other types into the urlencoder. I guess the urlencoder itself handles numbers. In short I don't think this change is risky in the behaviour change from passing `str` instead of e.g. `int` in the request `params` dict.